lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -395,56 +438,255 @@ void LoopTreeOptimization::propagateSummaries(
   }
 }
 
+static bool isSafeReadOnlyApply(SideEffectAnalysis *SEA, ApplyInst *AI) {
+  FunctionSideEffects E;
+  SEA->getCalleeEffects(E, AI);
+
+  if (E.getGlobalEffects().mayRead()) {
+    // If we have Global effects,
+    // we don't know which memory is read in the callee.
+    // Therefore we bail for safety
+    return false;
+  }
+
+  auto MB = E.getMemBehavior(RetainObserveKind::ObserveRetains);
+  return (MB <= SILInstruction::MemoryBehavior::MayRead);
+}
+
+static void checkSideEffects(swift::SILInstruction &Inst, WriteSet &MayWrites) {
+  if (Inst.mayHaveSideEffects()) {
+    MayWrites.insert(&Inst);
+  }
+}
+
+/// Returns true if the \p Inst follows the default hoisting heuristic
+static bool canHoistUpDefault(SILInstruction *inst, SILLoop *Loop,
+                              DominanceInfo *DT, bool RunsOnHighLevelSil) {
+  auto Preheader = Loop->getLoopPreheader();
+  if (!Preheader) {
+    return false;
+  }
+
+  if (isa<TermInst>(inst) || isa<AllocationInst>(inst) ||
+      isa<DeallocationInst>(inst)) {
+    return false;
+  }
+
+  if (inst->getMemoryBehavior() == SILInstruction::MemoryBehavior::None) {
+    return true;
+  }
+
+  if (!RunsOnHighLevelSil) {
+    return false;
+  }
+
+  ArraySemanticsCall semCall(inst);
+  return semCall.canHoist(Preheader->getTerminator(), DT);
+}
+
+static void analyzeBeginAccess(BeginAccessInst *&BI,
+                               SmallVector<BeginAccessInst *, 8> &BeginAccesses,
+                               SmallVector<EndAccessInst *, 8> &EndAccesses,
+                               HoistPairSet &HoistingPairs,
+                               const FunctionAccessedStorage &functionAccess) {
+  if (BI->getEnforcement() != SILAccessEnforcement::Dynamic) {
+    return;
+  }
+
+  const AccessedStorage &storage =
+      findAccessedStorageNonNested(BI->getSource());
+  if (!storage) {
+    return;
+  }
+
+  // find matching end access:
+  auto matchingEndPred = [&](EndAccessInst *EI) {
+    return EI->getBeginAccess() == BI;
+  };
+  auto matchingEnd =
+      std::find_if(EndAccesses.begin(), EndAccesses.end(), matchingEndPred);
+  if (matchingEnd == EndAccesses.end()) {
+    // no matching end within the loop
+    return;
+  }
+  auto *EI = *matchingEnd;
+  ++matchingEnd;
+  if (std::find_if(matchingEnd, EndAccesses.end(), matchingEndPred) !=
+      EndAccesses.end()) {
+    // We expect a single matching end access in the loop
+    return;
+  }
+
+  if (functionAccess.hasUnidentifiedAccess()) {
+    // safe mode - maybe we can still do something
+    if (BeginAccesses.size() > 1) {
+      // Too risky
+      return;
+    }
+    if (!BI->hasNoNestedConflict()) {
+      return;
+    }
+    HoistingPairs.push_back(std::make_pair(BI, EI));
+    return;
+  }
+
+  const StorageAccessInfo &storageInfo =
+      functionAccess.getStorageAccessInfo(storage);
+  if (!storageInfo.hasNoNestedConflict()) {
```

Comment:
I think you only care about the current instance of the access. So `AccessedStorageAnalysis.hasNoNestedConflict` is too strict. Since we don't care about other accesses to the same storage, you could just check the `[no_nested_conflict]` flag on `begin_access`.